src/cluster/placement/selector/mirrored_custom_groups.go

```diff
@@ -0,0 +1,238 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package selector
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3/src/cluster/placement"
+	"github.com/m3db/m3/src/x/errors"
+
+	"go.uber.org/zap"
+)
+
+type mirroredCustomGroupSelector struct {
+	instanceIDToGroupID InstanceGroupIDFunc
+	logger              *zap.Logger
+	opts            MirroredCustomGroupOptions
+}
+
+// MirroredCustomGroupOptions configure the selector from NewMirroredCustomGroupSelector
+type MirroredCustomGroupOptions interface {
+	// ZoneOptions configure the zone related options of the selector
+	ZoneOptions() placement.ZoneOptions
+
+	// SetZoneOptions -- see ZoneOptions
+	SetZoneOptions(zoneOpts placement.ZoneOptions) MirroredCustomGroupOptions
+}
+
+// NewMirroredCustomGroupOptions returns default options for a MirroredCustomGroupSelector.
+func NewMirroredCustomGroupOptions() MirroredCustomGroupOptions {
+	return mirroredCustomGroupOptions{
+		zoneOpts: placement.NewZoneOptions(),
+	}
+}
+
+type mirroredCustomGroupOptions struct {
+	zoneOpts placement.ZoneOptions
+}
+
+func (m mirroredCustomGroupOptions) ZoneOptions() placement.ZoneOptions {
+	return m.zoneOpts
+}
+
+func (m mirroredCustomGroupOptions) SetZoneOptions(
+	zoneOpts placement.ZoneOptions,
+) MirroredCustomGroupOptions {
+	m.zoneOpts = zoneOpts
+	return m
+}
+
+// InstanceGroupIDFunc maps an instance to its mirrored group.
+type InstanceGroupIDFunc func(inst placement.Instance) (string, error)
+
+// NewMapInstanceGroupIDFunc creates a simple lookup function for an instances group, which
+// looks up the group ID for an instance by the instance ID.
+func NewMapInstanceGroupIDFunc(instanceToGroup map[string]string) InstanceGroupIDFunc {
+	return func(inst placement.Instance) (string, error) {
+		gid, ok := instanceToGroup[inst.ID()]
+		if !ok {
+			return "", fmt.Errorf(
+				"instance %s doesn't have a corresponding group in ID to group map",
+				inst.ID(),
+			)
+		}
+		return gid, nil
+	}
+}
+
+// NewMirroredCustomGroupSelector constructs a placement.InstanceSelector which assigns shardsets
+// according to their group ID (provided by instanceToGroupID). That is, instances with the
+// same group ID are assigned the same shardset ID, and will receive the same mirrored traffic.
+func NewMirroredCustomGroupSelector(
+	logger *zap.Logger,
+	instanceToGroupID InstanceGroupIDFunc,
```

Comment:
could move `InstanceGroupIDFunc` & logger to the options to simplify the ctor (and it's the pattern used in most of this repo)

Reply:
logger I'm with you for sure since it's optional; `instanceToGroupID` is required, so I'm inclined to give the user a heads up about that. Reasonable in between? I really hate constructing these classes and having to wade through nil pointers for 30 minutes haha.

Reply:
fair enough